include/grpcpp/server_builder.h

```diff
@@ -210,15 +210,43 @@ class ServerBuilder {
   /// doc/workarounds.md.
   ServerBuilder& EnableWorkaround(grpc_workaround_list id);
 
- private:
-  friend class ::grpc::testing::ServerBuilderPluginTest;
-
+ protected:
   struct Port {
     grpc::string addr;
     std::shared_ptr<ServerCredentials> creds;
     int* selected_port;
   };
 
+  typedef std::unique_ptr<grpc::string> HostString;
+  struct NamedService {
+    explicit NamedService(Service* s) : service(s) {}
+    NamedService(const grpc::string& h, Service* s)
+        : host(new grpc::string(h)), service(s) {}
+    HostString host;
+    Service* service;
+  };
+
+  std::vector<Port> ports() { return ports_; }
+
+  std::vector<std::reference_wrapper<NamedService>> services() {
```

Comment:
I still don't like this. Sorry.

Reply:
This has just become API. Needs a gRFC or mark it experimental if you intend to deprecate it later.